### PR TITLE
Use Pydantic for type annotations and FastAPI/OpenAPI for HTTP API

### LINKED ISSUE #58: Default values are sometimes the string "None"

### LINKED ISSUE #113: Define path types without having to import pathlib
This is clumsy to explain in getting started, and means we have to have extra imports in all the model definition docs. Also a thing users will stumble on.

### LINKED ISSUE #193: Revisit design of Python type annotations
`@cog.input()` is separate from the actual type signature and involves referencing arguments with strings. We are also doing a bunch of inventing-here with type checking and API generation.

Can we do something more elegant with Python types/Pydantic/FastAPI?

## Blockers

- Probably depends on #259 to ensure we can output correct APIs.
- Closely intwined with #205 -- they might be one and the same thing if we use something like FastAPI.

## Proposed design

Pydantic and [FastAPI](https://fastapi.tiangolo.com/tutorial/query-params-str-validations/) have a pattern where the default value can be an object with more details about the object.

For example, to translate an example from [our docs](https://github.com/replicate/cog/blob/main/docs/getting-started-own-model.md):

```python
import cog
from pathlib import Path
import torch


class Predictor(cog.Predictor):
    def setup(self):
        self.net = torch.load("weights.pth")

    def predict(
        self,
        input: Path = cog.Input(description="Image to enlarge"),
        scale: float = cog.Input(default=1.5, description="Factor to scale image by"),
    ) -> Path:
        """Run a single prediction on the model"""
        # ... pre-processing ...
        output = self.net(input)
        # ... post-processing ...
        return output

```

`cog.Input` is the same as `pydantic.Field`. We could just use that.

The beauty of incorporating Python types and FastAPI is:
- IDE type completion / checking works
- We can automatically generate a validating API and an OpenAPI schema for the model (which has a number of additional benefits: automatic docs, automatic forms, automatic client libraries, etc)

### LINKED ISSUE #205: Design type system & signature
The type system and signature needs some thought.

## User data

- An industry user wants a way of documenting the inputs and outputs of their models so they can figure out how they fit together. Their models use a combination of video, image, and text content with additional metadata that could be represented as JSON.
- We, Replicate, need a type signature to generate forms on Replicate. Specific problems we have encountered:
  - ints and bools are strings
  - There is no image type
  - There are no output types

## Requirements

- We need to define input and output types for a prediction function. It should have some kind of schema (what we currently have at `GET /type-signature`).
- Standard types and custom types (e.g. images, or complex objects) need to be represented.
- Multiple or complex outputs need to be represented somehow. #212 
- It needs to be able to represent files.
- It should support standard REST JSON + HTTP, plus other representations and transports (e.g. gRPC, AMQP)
- Ideally, we want a standard off-the-shelf system (e.g. OpenAPI).

## Potential implementations

We could use OpenAPI / JSON Schema, which would handily generate a self-documenting API, instead of our unnecessarily-invented-here `/type-signature` endpoint. This might happen automatically if we make use of FastAPI in #193. (Hello @zeke!)

A potential issue with this is mixing in files to input/output. There are a few potential solutions: gRPC, multipart forms, references to external files (but how do output files work?). Maybe there is prior art here we can learn from.

## See also

- #193 
- #259 (blocked on / closely related)

### LINKED ISSUE #212: Output types and multiple outputs
We support multiple inputs and types for them. We should do the same for outputs.

Probably depends on #205

### LINKED ISSUE #246: Cog requires inputs for predict function
There's no reason that it should. This should work:

```
class Predictor(cog.Predictor):
    def setup(self):
        # ...

    def predict(self):
        return "hello"
```

It at least should not throw a confusing error message about "`_inputs` not set" or something.

### LINKED ISSUE #259: Review model API design
The current model API was hastily written, and we have now learned many things that could be incorporated into its design.

## User data

- Replicate! We currently use an undocumented Redis queue consumer.
- An industry user wants a gRPC API with support for non-blocking requests for long-running models.
- An industry user wants an AMQP RPC API. It is unclear whether the model itself needs to interact over this API, or whether a sidecar 

### Essential

- The primary API for Cog should be the same thing we use for Replicate.
- Make predictions as standard blocking REST API. JSON over HTTP.
- Make it possible to use with a queuing system. Either built-in, as a sidecar at the container-level, or as an extension at the Python-level.
- Pass files as either URLs or base64 encoding, depending on if efficiency is important. (Interdependent with #205.)

### Future, to design in context

- gRPC interface. This is harder to use, so shouldn't be the primary interface, but it seems to be becoming widely used for ML.
- Other serving platforms, like AI Platform or Kserve.
- Non-blocking API with requests
- #230 

## Prior art

We don't need to reinvent the wheel.

### Off the shelf

- https://www.tensorflow.org/tfx/serving/api_rest
- Kserve's model API: https://github.com/kserve/kserve/blob/master/docs/predict-api/v2/required_api.md
- Kserve has a sidecar queue proxy: https://github.com/kserve/website/blob/main/docs/modelserving/control_plane.md
- Sagemaker? (Can anyone find the actual API documentation?)
- https://github.com/triton-inference-server/server/blob/main/docs/architecture.md#models-and-schedulers
- https://pytorch.org/serve/rest_api.html

### Real world

- How is it done at spotify @andreasjansson?
- Might be some descriptions of how people do serving here: https://huyenchip.com/mlops/

## Areas for discussion

- What are the trade-offs between an HTTP based API and a queue based API?
- How do we support queuing systems?

## Next steps
- [ ] Use default Cog server for Replicate

## See also
- #193
- #205 
- #212

(Maintainers -- please edit this and consider it a wiki! Edited by @bfirsh, ...)

### LINKED ISSUE #262: Integration tests should use Docker networking instead of relying on host's networking
The redis integration tests connect to redis by going out of Docker, to the host machine, then back into Docker via the exposed port, using the IP address of the host machine.

This is fragile because it might break depending on the network setup of the host machine. A more reliable way would be to use Docker networks (and just connect to `redis:6379` inside the model container).

### LINKED ISSUE #304: Everything in predict method but `name` needs to be a named argument
When attempting to specify `type` and `help` without names:

```py
@cog.input("seed", str, "A random seed", min=0, max=1000, default=None)
```

The following error is produced:

```
E   TypeError: input() got multiple values for argument 'default'
```

### LINKED ISSUE #306: String max_length parameter on @cog.input
We don't have validation on string length in @cog.input. A `max_length` field could provide that.

Additionally it could be used when building UIs on top of Cog, e.g. Replicate.ai, to determine if the input element should be a single line input or multi-line textbox.

### LINKED ISSUE #327: Strawman: FastAPI model server
This is potential design that could hit three interconnected birds with one stone:
- 🐣 #193 
- 🐥 #205 
- 🐔 #259 

Those tickets contain user data and requirements. Please read them for context.

This is a strawman. It is most likely wrong. Please attack with your sharpest pitchforks!

## Model definition

The model API is specified using Pydantic types in the predict function.

```python
import cog
from pathlib import Path
from pydantic import Field
import torch


class Predictor(cog.Predictor):
    def setup(self):
        self.net = torch.load("weights.pth")

    def predict(
        self,
        input: Path = Field(..., title="Image to enlarge"),
        scale: float = Field(1.5, description="Factor to scale image by"),
    ) -> Path:
        """Run a single prediction on the model"""
        # ... pre-processing ...
        output = self.net(input)
        # ... post-processing ...
        return output
```

You could imagine some custom types like `cog.Field` if we need custom behavior, or don't want the extra import.

This predictor serves:

### `GET /predict`

Takes a JSON object in request body:

```json
{
  "input": "https://bucket.s3.amazon.com/...", # or "data:foo/bar;base64,..."
  "scale": 1.7,
}
```

Returns:

```
{
  "output": "https://bucket.s3.amazon.com/..."
}
```

Everything is JSON. Files are represented as URLs -- HTTP, cloud storage, data: URLs, or perhaps even file:// for mounted storage. This is for simplicity, so it can be represented sensibly in OpenAPI, and so more complex data structures can be represented.

If a more complex Pydantic object was defined as the return value, `output` would be a JSON object. The root object can also include other information about the run -- some prior art in #259 might be informative about what we want to include (metrics, etc).

Authentication and configuration for cloud storage needs some thought. It could specify a prefix for storing output files in the request.

### `GET /docs`

A Swagger UI, generated automatically by FastAPI.

### `GET /openapi.json`

An OpenAPI definition, generated automatically by FastAPI. This is used as the type definition. It contains a complete documentation of the HTTP API and entries like this to define the types for input/output:

```
{
  "components": {
    "schemas": {
      "Input": {
        "title": "Input",
        "required": [
          "input"
        ],
        "type": "object",
        "properties": {
          "input": {
            "title": "Image to enlarge",
            "type": "string",
            "format": "path"
          },
          "scale": {
            "title": "Scale",
            "type": "number",
            "description": "Factor to scale image by",
            "default": 1.5
          }
        }
      },
      "Output": ...
    }
  }
}
```

## How this would be used for Replicate

We could either:

1. Integrate with existing redis queue worker.
2. Run a sidecar container that converts the current Redis API into HTTP.

Over time, we could define a queue API and include that in Cog. (Or leave it as an external system. See discussion in #259.)

## Files

Files are represented as URLs. `data:`, `http:`, `gcs:`, `s3:`, etc.

This works well for input, but some way of specifying where output files go is required. Predictions could produce arbitrary output files (or should they be able to?) so some generic way of defining where multiple files can go is needed. For filesystem or bucket, this could be a prefix, and the prediction is responsible for naming files such that they don't collide.

Perhaps prediction requests could include the output path prefix:

```
{
  "input": {"image": "s3://predictions/1/input/image.jpg"},
  "output_path": "s3://predictions/1/output/",
}
```

~For Replicate, files could be fetched and uploaded from Cloud Storage using signed URLs with a prefix.~ Unfortunately signed URLs can't be made for prefixes. You can assign IAM permissions for a prefix, but that would mean a service account for each prediction, which is unwieldy.

Some potential solutions:

1. We use serving's existing file storage system
2. We send files directly back to replicate-web, bypassing serving's file storage system. Perhaps if 
3. We use data URIs...?!

## Future

- An asynchronous API, where a job is started and you can query the status of the job. (This might be required for passing logs to Replicate?)
- This request/response can be passed over a queue. The same type definition could be used.
- How would do we support gRPC? Would we want to generate a generic protobuf schema that supports any kind of model, or generate one per-model somehow?
- We might want a more FastAPI-like API instead of the class-based `Predictor`.
- Custom URL routes on the FastAPI server. These might be namespaced under `/extensions` to preserve the stability of the top-level API.

## Reflection

FastAPI is very HTTP + JSON centric. OpenAPI is too. This is optimal for simple use-cases, but using more advanced/efficient systems like gRPC or queues may feel like an impedance mismatch.

## Implementation details

The predictor in the example above would be converted into a FastAPI method like this:

```python
app = FastAPI()

class Input(BaseModel):
    input = Path = Field(..., title="Image to enlarge")
    scale = float = Field(1.5, description="Factor to scale image by")

@app.post("/predict")
async def predict(input: Input):
    # ...
    return Path(...)
```

We would need to come up with a custom `Path` handler that knows how to convert it into URLs or base64 encoded strings.

### PR DESCRIPTION
**Pull request by bfirsh**
The exploration in #343 was sufficiently explored and getting very messy, so I thought I would create a new pull request for the real implementation.

This is an implementation of #193, #205, and #259 using Pydantic for the type system, FastAPI for the model server, and OpenAPI for the schema.

Models now look a bit like this:

```python
from cog import Predictor, Input, Path
import torch

class ColorizationPredictor(Predictor):
    def setup(self):
        self.model = torch.load("./weights.pth")

    def predict(self,
          image: Path = Input(title="Grayscale input image")
    ) -> Path:
        processed_input = preprocess(image)
        output = self.model(processed_input)
        return postprocess(output)
```

The HTTP API is:

- `POST /predictions`: Run a prediction.
- `GET /openapi.json`: OpenAPI schema for HTTP API, including input and output JSON Schema types generated from the `predict()` type signature.

Files are passed as URLs. Currently only `data:` and `http:` URLs are supported. It is intended that `file:`, `http:`, `s3:`, and `gs:` URLs will be supported. For Replicate, we will pass files using HTTP directly from web. Output files will be POSTed directly back to web.

The Redis queue worker remains in place. It now uses the same request and response objects as the HTTP API. The tests should explain its behavior. Logging still uses the old logging API, but is intended to add this to the response object like the Replicate serving API.

Commits are significant and this should be reviewed commit by commit. This is an intentionally large and long-running branch -- `main` is currently used as the documentation for Replicate.

## Todo
- [x] `cog predict` with new API
- [x] Input ordering?
- [x] Ensure HTTP server is single threaded.
- [x] HTTP URLs for input and output
- [x] Add `choices` option to `Input` to generate an `Enum`.
- [x] `cog init` template
- [x] Update getting started docs
- [x] Update reference docs
- [x] Rename `Predictor` to `BasePredictor` so we can do `from cog import BasePredictor, Input, Path` without colliding with a `Predictor` subclass.
- [x] Better errors if you get output type wrong.
- [x] ~Reliably set Cog version. In development, the version is set to `0.0.1` which is pretty useless. It should probably be hardcoded somewhere, and the release process needs updating. We need this to be able to switch between old and new Cog API in Replicate.~ We're just going to support version `dev` everywhere which means "latest".
- [x] Update examples https://github.com/replicate/cog-examples/pull/8
- [x] Update the OpenAPI Outputs to be a consistent place. (Simple outputs currently end up in Request object schema)
- [x] Test returning object as output
- [x] Test arbitrary JSON objects in output
- [ ] Is `go.sum` up to date?

## Future (Pydantic + OpenAPI)
- [ ] Tell users to upgrade if they use `@cog.input()` with new Cog.
- [ ] Document nouns to make sure we're all on the same page. "Input"/"Inputs"?
- [ ] Document how to migrate `@input()` to Pydantic type annotations.
- [ ] Go through and fix any TODOs
- [ ] Make input validation errors work for `cog predict` in a simple way until we lock down HTTP API
- [ ] Make input errors work for Redis API & Replicate
- [ ] Integrate with Replicate - https://github.com/replicate/replicate-web/pull/907
- [ ] Force users to pick an output type -- throw an error if it's not set. (Couldn't just be `Any`, but they need to be explicit!)
- [ ] Clean up temporary files -- both input and output
- [ ] Do we want to make `cog predict` with an existing image backwards compatible? ('cos there are old models on Replicate)
- [ ] What happens if you use `pathlib.Path` as an input type? Does it work? Should we throw an error? (This might happen if you import wrong package / collide imports)
- [ ] What happens if you return `pathlib.Path` from a predict function? Does this work? Should it throw an error?
- [ ] Should we have a fixed set of supported inputs/outputs somehow?
- [ ] Do we want to force users to put extra properties in `additionalProperties`? Should we even allow extra properties to `cog.Input()`?
- [ ] Merge examples PR https://github.com/replicate/cog-examples/pull/8

## Far-flung future
- Vendor Pydantic and FastAPI
- Add back support for PyTorch and Tensorflow tensors as output. This was removed because it was fiddly to get working with Pydantic. You now need to convert them to numpy arrays.
- Add back timing information to server.
- Better error if you forget an output type
    ```
    pydantic.error_wrappers.ValidationError: 1 validation error for Prediction
    output
      unexpected value; permitted: None (type=value_error.const; given=hello foo; permitted=(None,))
    ```
- Multiple file outputs for `cog predict`. It currently just outputs the plain JSON if there isn't a single file output.
- Logging, as described in comment in #343
- Review request/response objects and make sure we're happy with them.
     - This might be an opportunity to rename the statuses and ensure they're consistent ("success" vs "failed").
     - Do we want the input type to have a second `input` level so we can have top-level request options?
- Do we need separate URLs for input/output schema, or can they be reliably fetched from OpenAPI?
- [ ] Document HTTP API
- [ ] Do we want to version the prediction API, or let the client handle it? https://github.com/replicate/cog/issues/94
- [ ] Test arbitrary objects in input
- [ ] What if people run new models from a directory of code with old Cog? What happens if we do that? Do we care? #286 
- [ ] What if people run new models from an image with old Cog? Maybe we throw a simple error from `/predict`?

## Based on (& blocked by)

- #383
- #385 

## Refs
Closes #58
Closes #113 
Closes #193
Closes #205
Closes #212 
Closes #246
Closes #259
Closes #262
Closes #263
Closes #304 
Closes #306 
Closes #327
Closes #328

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,16 +1,14 @@
 # Cog: Containers for machine learning
 
-Use Docker for machine learning, without all the pain.
+Put your machine learning model in a standard, production-ready Docker container without having to know how Docker works.
```
Comment (zeke) [inline README.md:3]: ❤️

File: README.md [hunk 2]
```diff
@@ -87,10 +63,34 @@ $ cog build -t my-colorization-model
 
 $ docker run -d -p 5000:5000 --gpus all my-colorization-model
 
-$ curl http://localhost:5000/predict -X POST -F input=@image.png
+$ curl http://localhost:5000/predictions -X POST \
+   --data '{"input": "https://.../input.jpg"}'
 ```
 
-That's it! Your model will now run forever in this reproducible Docker environment.
+<!-- In development, you can also run arbitrary commands inside the Docker environment:
```
Comment (zeke) [inline README.md:70]: Why is this commented out?

File: README.md [hunk 3]
```diff
@@ -87,10 +63,34 @@ $ cog build -t my-colorization-model
...
+```
+$ cog run jupyter notebook
+```
+-->
+
+## Deploying models to production
+
+Cog does a number of things out of the box to help you deploy models to production:
+
+- **Standard interface.** Put models inside Cog containers, and they'll run anywhere that runs Cog containers.
```
Comment (zeke) [inline README.md:89]: `they'll run anywhere that runs Cog containers.` --- Should this be **Docker** instead of **Cog**

Comment (bfirsh) [reply README.md:89]: I mean, technically correct, but yes that is a more useful statement haha.

File: README.md (hunk 2 above)
Comment (bfirsh) [reply README.md:70]: Just tinkering with this. Perhaps it's more clear what Cog is for (packaging models into Docker containers) when there isn't the diversion into development environments.

File: docs/getting-started-own-model.md [hunk 4]
```diff
@@ -88,16 +88,25 @@ class Predictor(cog.Predictor):
...
+- `int`: an integer
+- `float`: a floating point number
+- `bool`: a boolean
+- `cog.File`: a file-like object representing a file
+- `cog.Path`: a path to a file on disk
+
+You can provide more information about the input with the `Input()` function, as shown above. It takes these basic arguments:
+
+- `description`: A description of what to pass to this input for users of the model
 - `default`: A default value to set the input to. If this argument is not passed, the input is required. If it is explicitly set to `None`, the input is optional.
```
[the remark this answers is not in the document]
Comment (bfirsh) [reply docs/getting-started-own-model.md:103]: There could be an argument for a more explicit, easier to understand design here. https://github.com/replicate/cog/issues/114 Not sure what is best.

**Comment by bfirsh** [comment]
I have rebased this on main to get the vscode editor goodness.

**Comment by bfirsh** [comment]
@zeke Looks like the docstrings do indeed end up in the schema, as the tests have told us. 😂

**Comment by zeke** [comment]
> docstrings do indeed end up in the schema

Gotcha. I'll take a look.

**Comment by zeke** [comment]
🍏

**Comment by bfirsh** [comment]
This is an intentionally large and long-running branch because `main` is used as the documentation for Cog.

I wonder if we could get this work into `main`, but not update the documentation? Or maybe in a separate long-running branch on this repository that we can all easily open pull requests against?

**Comment by zeke** [comment]
> I wonder if we could get this work into main, but not update the documentation? Or maybe in a separate long-running branch on this repository that we can all easily open pull requests against?

I'd be in favor of the latter: long-running branch on this repo (not a fork, and with a cool name)

1. Create a new branch from `main` called `new-cog` or `year-of-the-tiger`. Can we use emojis in branch names? 🤔 
1. Point this PR at `new-cog`
1. Merge this PR
1. Open followup PRs against `new-cog`

**Comment by bfirsh** [comment]
👍

To summarize an IRL conversion: we chose the extremely cool name "future" for the branch.

I have cleaned up the commits a little (squashed some "fix tests" commits etc), updated go.mod, and based this PR on our new future branch. We're ready to roll, I think!

Next steps:
- [ ] Create a new Mega Pull Request for the "future" branch. This branch will be an initial version of Cog that we can use to start testing it with Replicate internally.
- [ ] Migrate "Future (Pydantic + OpenAPI)" tasks to that pull request.
- [ ] Create issues for the "Far flung future" and add them to the "next" in the project to remind us to do them before we ship new cog.

**Review by zeke — APPROVED** [review]
The `future` looks bright. 😎 

Ship it (to a feature branch)! 🚢

**Comment by bfirsh** [comment]
In case it's not clear from the previous conversation: we decided to defer the design of the HTTP API to the future. The design in the body of this pull request is merely a strawman for a future design that we will lock down.

**Comment by bfirsh** [comment]
I have added a few tickets for the far flung future that can be seen in the mentions above. They're in the "next" status in our project to remind us to look at them before we ship new Cog.

**Pull request merged**